core/src/util/transactional_hashmap.rs

```diff
@@ -142,18 +146,14 @@ where
 	V: Clone,
 {
 	fn commit(&mut self) {
-		self.rollback_operations = vec![];
+		self.original_items = HashMap::new();
 	}
 
 	fn rollback(&mut self) {
-		while !self.rollback_operations.is_empty() {
-			let op = self.rollback_operations.pop().unwrap();
-			match op {
-				Reversible::Add { prev, key } => match prev {
-					Some(old) => self.inner.insert(key, old),
-					None => self.inner.remove(&key),
-				},
-				Reversible::Remove { prev, key } => self.inner.insert(key, prev),
+		for (key, v) in self.original_items.drain() {
+			match v {
+				Some(old) => self.inner.insert(key, old),
+				None => self.inner.remove(&key),
```

Comment:
Why not just:
```
self.inner = self.original_items;
self.original_items = HashMap::new();
```

Reply:
we can't, originals only keeps track of the originals that are overridden. Maybe I need to rename it to `overriden_originals` and inner is the latest state for all items.